awx/main/scheduler/dag_simple.py

```diff
@@ -122,3 +120,30 @@ def get_root_nodes(self):
             if len(self.get_dependents(n['node_object'])) < 1:
                 roots.append(n)
         return roots
+
+    def has_cycle(self):
+        node_objs = [node['node_object'] for node in self.get_root_nodes()]
+        node_objs_visited = set([])
+        path = set([])
+        stack = node_objs
+
+        if len(self.nodes) != 0 and len(node_objs) == 0:
+            return True
+
+        while stack:
+            node_obj = stack.pop()
+
+            children = [node['node_object'] for node in self.get_dependencies(node_obj)]
+            children_to_add = filter(lambda node_obj: node_obj not in node_objs_visited, children)
```

Comment:
This seems inefficient in the following sense.

Visited children are not added onto the stack by this clause, but that's only using a definition of "visited" in that the node has been the `node_obj` in the past. It's still possible that an entry in `children_to_add` already exists as a prior node in `stack`.

Consider...

root node A has children B and C. The way this loop crawls is that `stack` starts as `[A]`, gets shortened to `[]`, has it added back on, `[A]`, has children added `[A,B,C]`, and then proceeds to the next loop.

In next loop, `node_obj` is C and `stack` is `[A,B]`. Now, here is where I throw the gotcha. C has children B and D. Since B was never processed as a `node_obj` by this point (it needs to reverse back up through `stack`), it gets added on again. The list becomes `[A,B,B,D]`.

You could then argue some kind of extreme theoretical case. Say that A had children B through Z, plus alpha. Alpha also has children B through Z, plus beta. Beta has children B through Z, plus gamma. Continue infinitely, and `stack` gets spammed with repeated entries.